services/matching-service/start.ts

```diff
@@ -0,0 +1,120 @@
+import express from "express";
+import { Server } from "socket.io";
+import { createServer } from "http";
+
+import { extractCategory, extractDifficulty, isDuplicateID as checkIdExists } from "./domain/match_validators";
+
+const app = express();
+const httpServer = createServer(app);
+const io = new Server(httpServer);
+
+
+const port = process.env.PORT || 6001;
+
+io.on("connection", (socket) => {
+    socket.emit("reply", "hello from server");
+
+    socket.on("register", (msg) => {
+        console.log(`A NEW USER WITH ID ${msg.id} JOINED!`);    
+    
+        const difficulty = extractDifficulty(msg.difficulty);
+        const category = extractCategory(msg.category);
+        
+        // Check Valid Difficulty
+        if (difficulty == undefined) {
+            socket.emit("error", "Invalid Difficulty!");
+            return;
+        }
+        
+        //Check valid Category
+        if (category == undefined) {
+            socket.emit("error", "Invalid Category!");
+            return;
+        }
+        
+        // Check for duplicate ID and push to Queue
+        if (checkIdExists(msg.id)) {
+            socket.emit("error", "You are already in the queue!");
+            return;
+        } 
+
+        // Find match if exists, else create new queue and wait.
+        const this_match = {
```

Comment:
I've kept this on top because we might need to add the user to the queue even if it is not empty (i.e the case when the user's peer was added to the queue, but his connection was lost and peerSocket is undefined).